velox/functions/lib/Re2Functions.cpp

```diff
@@ -970,89 +970,266 @@ std::vector<std::shared_ptr<exec::FunctionSignature>> re2ExtractSignatures() {
   };
 }
 
-PatternMetadata determinePatternKind(StringView pattern) {
-  vector_size_t patternLength = pattern.size();
-  vector_size_t i = 0;
-  // Index of the first % or _ character.
-  vector_size_t wildcardStart = -1;
-  // Count of wildcard character sequences in pattern.
-  vector_size_t numWildcardSequences = 0;
+std::string unescape(
+    StringView pattern,
+    size_t start,
+    size_t end,
+    std::optional<char> escapeChar) {
+  if (!escapeChar) {
+    return std::string(pattern.data(), start, end - start);
+  }
+
+  std::ostringstream os;
+  auto cursor = pattern.begin() + start;
+  auto endCursor = pattern.begin() + end;
+  while (cursor < endCursor) {
+    auto previous = cursor;
+
+    // Find the next escape char.
+    cursor = std::find(cursor, endCursor, escapeChar.value());
+    if (cursor < endCursor) {
+      // There are non-escape chars, append them.
+      if (previous < cursor) {
+        os.write(previous, cursor - previous);
+      }
+
+      // Make sure there is a following normal char.
+      VELOX_USER_CHECK(
+          cursor + 1 < endCursor,
+          "Escape character must be followed by '%', '_' or the escape character itself");
+
+      // Make sure the escaped char is valid.
+      cursor++;
+      auto current = *cursor;
+      VELOX_USER_CHECK(
+          current == escapeChar || current == '_' || current == '%',
+          "Escape character must be followed by '%', '_' or the escape character itself");
+
+      // Append the escaped char.
+      os << current;
+    } else {
+      // Escape char not found, append all the non-escape chars.
+      os.write(previous, endCursor - previous);
+      break;
+    }
+
+    // Advance the cursor.
+    cursor++;
+  }
+
+  return os.str();
+}
+
+// An iterator to iterate through a pattern string. It will handle escaping
```

Comment:
nit: "It will handle" -> "Handles escaping transparently"

or

`// Iterates through a pattern string. Transparently handles escape sequences.`